src/Database/Esqueleto/ForeignKey.hs

```diff
@@ -0,0 +1,245 @@
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TupleSections #-}
+
+module Database.Esqueleto.ForeignKey
+  ( ForeignPair
+  , ForeignKey
+  , mkForeignInstances
+  , onForeignKey
+  , foreignKey
+  , foreignKeyL
+  , foreignKeyR
+  , foreignKeyLR
+  , foreignKeyLMaybe
+  , foreignKeyRMaybe
+  , foreignKeyLRMaybe
+  )
+
+where
+
+import qualified Data.Char                                    as Char
+import qualified Data.Function                                as Function
+import qualified Data.List                                    as List
+import           Data.Monoid
+import qualified Data.Ord                                     as Ord
+import qualified Data.Text                                    as Text
+import           Data.Traversable                             (forM)
+import qualified Language.Haskell.TH                          as TH
+
+import           Database.Esqueleto.Internal.Language
+import           Database.Esqueleto.Internal.PersistentImport
+
+-- | Describe a Pair of keys that are part of a foreign key relationship.
+--
+-- The first element is the entity field that holds the foreign key. The second
+-- element is the entity field that holds the references primary key. Type of
+-- primary key and foreign key have to coincide.
+--
+-- For example, give the following Entity definition:
+--
+-- @
+-- Employee
+--     num Int
+--     Primary numid
+--     name Text
+--
+-- Team
+--     teamId Int
+--     employee Int
+--     Foreign Employee fkEmployee employee
+-- @
+--
+-- The Following ForeignPair would capture the foreign key relationship
+--
+-- @ForeignPair TeamEmployee EmployeeNum@
+
+data ForeignPair :: * -> * -> * where
+    ForeignPair :: forall a b f.
+                   (PersistEntity a, PersistEntity b, PersistField f) =>
+                    EntityField a f
+                 -> EntityField b f
+                 -> ForeignPair a b
+
+-- | Describe a unique, canonical foreign key relationship between entities. For
```

Comment:
Yes, that restriction is pretty unfortunate. An alternative would be to create named values instead and refer to them. The cost is that now we have to remember the name of the foreign-key relationship (instead of the name of the fields), which the typeclass approach avoids. Persistent does indeed store the foreign key information in various forms, but at least as far as I can see only as `Text`s.

Reply:
>The cost is that now we have to remember the name of the foreign-key relationship

If I am not mistaken, isn't this the case with how SQL database work anyway? You have to name your unique constraints in the support Persistent offers for this already, so it's not a departure from how Persistent works either.

It seems like you could generate sum type of valid foreign key relationships tied to a model to have well-typed references to FKs that aren't stringly. Similar to Yesod routing perhaps if you're familiar with that. I don't know what the Persistent representation looks like, that might be more suitable.